pystiche/papers/ulyanov_et_al_2016/loss.py

```diff
@@ -0,0 +1,172 @@
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    multi_layer_encoder: Optional[MultiLayerEncoder] = None,
+    layer: str = "relu_4_2",
+    score_weight=None,
+):
+    if score_weight is None:
+        if impl_params:
+            score_weight = 1e0 if instance_norm else 6e-1
+        else:
+            score_weight = 1e0
+
+    if multi_layer_encoder is None:
+        multi_layer_encoder = ulyanov_et_al_2016_multi_layer_encoder()
+    encoder = multi_layer_encoder[layer]
+
+    return UlyanovEtAl2016MSEEncodingOperator(encoder, score_weight=score_weight, impl_params=impl_params)
+
+
+class UlyanovEtAl2016GramOperator(GramOperator):
+    def __init__(
+        self, encoder: Encoder, impl_params: bool = True, **gram_op_kwargs: Any,
+    ):
+        super().__init__(encoder, **gram_op_kwargs)
+        self.normalize_by_num_channels = impl_params
+        self.double_batch_size_mean = impl_params
+
+    def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
+        gram_matrix = super().enc_to_repr(enc)
+        if not self.normalize_by_num_channels:
+            return gram_matrix
+
+        num_channels = gram_matrix.size()[-1]
+        return gram_matrix / num_channels
+
+    def calculate_score(
+        self, input_repr: torch.Tensor, target_repr: torch.Tensor, ctx: None
+    ) -> torch.Tensor:
+        score = super().calculate_score(input_repr, target_repr, ctx=ctx)
+        if not self.double_batch_size_mean:
+            return score
+        else:
+            batch_size = input_repr.size()[0]
+            return score / batch_size
+
+def ulyanov_et_al_2016_style_loss(
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    stylization: bool = True,
+    multi_layer_encoder: Optional[MultiLayerEncoder] = None,
+    layers: Optional[Sequence[str]] = None,
+    layer_weights: Union[str, Sequence[float]] = "sum",
+    score_weight: float = None,
+    **gram_op_kwargs,
+):
+    if score_weight is None:
+        if impl_params:
+            if instance_norm:
+                score_weight = 1e0
+            else:
+                score_weight = 1e3 if stylization else 1e0
+        else:
+            score_weight = 1e0
+
+    if multi_layer_encoder is None:
+        multi_layer_encoder = ulyanov_et_al_2016_multi_layer_encoder()
+
+    if layers is None:
+        if impl_params and instance_norm:
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1")
+        else:
+            layers = ("relu_1_1", "relu_2_1", "relu_3_1", "relu_4_1", "relu_5_1")
+
+    def get_encoding_op(encoder, layer_weight):
+        return UlyanovEtAl2016GramOperator(
+            encoder, score_weight=layer_weight, **gram_op_kwargs
+        )
+
+    return MultiLayerEncodingOperator(
+        multi_layer_encoder,
+        layers,
+        get_encoding_op,
+        layer_weights=layer_weights,
+        score_weight=score_weight,
+    )
+
+
+class UlyanovEtAl2016PerceptualLoss(MultiOperatorLoss):
```

Comment:
I've added a `PerceptualLoss` base class in #120. I think we can use it here. Please have a look.

Reply:
On a second thought, I don't think it will fit here. I designed it to always have a `content` and `style_loss` and optional `regularization`. Since the texture generation has no `content_loss` I think we should stick to the current implementation.

Reply:
Right I did not see that the `content_loss` here is not optional. However, wouldn't it make sense to design it that way? My thought is that it often happens that in the papers a distinction is made between `texture` and `stylization` and the `PerceptualLoss` in both situations only differ in the `content_loss` and the additional function `set_content_image`. I'm not sure, but the default `PerceptualLoss` for the `texture` will differ as little between papers as it does for `stylization`, correct?

Reply:
Yeah, you are right. I see two ways to implement this:

1. Make the `content_loss` optional in the current implementation or
2. Create an `StyleTransferPerceptualLoss` and `TextureSynthesisPerceptualLoss` for the two different cases.

What do you think?

Reply:
That's a good question. On the one hand, variant 1 is more flexible but not as intuitive, because you don't see directly that you can switch between `texture` and `stylization` with the optional `content_loss`. On the other hand, variant 2 is more intuitive, because you can see directly what the `PerceptualLoss` is for. However, an additional class/function is required that returns the respective `PerceptualLoss`. If you have no other arguments, I would prefer variant 2 for better readability. Maybe with a `default` class that can handle the switch between `texture` and `stylization` with the flag `stylization`? 

Regarding the optional `style_loss`: Is an `NST` without `style_loss` even possible? I have not seen a paper that doesn't use `style_loss`, which would make the optional inclusion necessary.

Reply:
> Is an NST without `style_loss` even possible?

Nope, but neither is one without `content_loss`. I would classify the different types like this

| type                   | `content_loss` | `style_loss` |
| ---------------------- | -------------- | ------------ |
| texture synthesis      | `False`        | `True`       |
| content reconstruction | `True`         | `False`      |
| NST                    | `True`         | `True`       |

The `regularization` is optional for every type. With that in mind I tend to option 1. If later someone complains about this being unintuitive, we can always build the special classes for every type on top.